src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -5,34 +5,650 @@
+        /// Similar in behavior to the x86 instruction POPCNT.
+        /// </summary>
+        /// <param name="value">The mask.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint PopCount(int value)
+            => PopCount(unchecked((uint)value));
+
+        /// <summary>
+        /// Returns the population count (number of bits set) of a mask.
+        /// Similar in behavior to the x86 instruction POPCNT.
+        /// </summary>
+        /// <param name="value">The mask.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint PopCount(ulong value)
+        {
+            if (Popcnt.IsSupported)
+            {
+                if (Popcnt.X64.IsSupported)
+                {
+                    return (uint)Popcnt.X64.PopCount(value);
+                }
+
+                // Use the 32-bit function twice
+                uint hv = (uint)(value >> 32); // High-32
+                uint bv = (uint)value; // Low-32
+
+                uint h = Popcnt.PopCount(hv);
+                uint b = Popcnt.PopCount(bv);
+
+                return h + b;
+            }
+
+            const ulong c0 = 0x_5555_5555_5555_5555;
+            const ulong c1 = 0x_3333_3333_3333_3333;
+            const ulong c2 = 0x_0F0F_0F0F_0F0F_0F0F;
+            const ulong c3 = 0x_0101_0101_0101_0101;
+
+            ulong val = value;
+
+            val -= (value >> 1) & c0;
+            val = (val & c1) + ((val >> 2) & c1);
+            val = (val + (val >> 4)) & c2;
+            val *= c3;
+            val >>= 56;
+
+            return (uint)val;
+        }
+
+        /* Legacy implementations
+        DONE https://raw.githubusercontent.com/dotnet/corefx/master/src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/BitArithmetic.cs
+        DONE https://raw.githubusercontent.com/dotnet/roslyn/367e08d8f9af968584d5bab84756eceda1587bd9/src/Workspaces/Core/Portable/Utilities/CompilerUtilities/ImmutableHashMap.cs
+        TBD https://raw.githubusercontent.com/dotnet/coreclr/030a3ea9b8dbeae89c90d34441d4d9a1cf4a7de6/tests/src/JIT/Performance/CodeQuality/V8/Crypto/Crypto.cs
+        */
+
+        #endregion
+
+        #region RotateRight
+
+        // Will compile to instrinsics if pattern complies (uint/ulong):
+        // https://github.com/dotnet/coreclr/pull/1830
+
+        /// <summary>
+        /// Rotates the specified value right by the specified number of bits.
+        /// Similar in behavior to the x86 instruction ROR.
+        /// </summary>
+        /// <param name="value">The value to rotate.</param>
+        /// <param name="offset">The number of bits to rotate by.
+        /// Any value outside the range [0..31] is treated as congruent mod 32.</param>
+        /// <returns>The rotated value.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint RotateRight(uint value, int offset)
```

Comment:
Existing callsites all use `uint` for offset, we should probably do that since negatives make no sense anyway

Reply:
I think it probably needs further discussion.

C# doesn't allow shifting by `uint`, so this at least keeps the semantics consistent with what users would have to write manually. However, there is also no reason why `uint` should be invalid (especially since C# masks the shift amount with 31).

Reply:
OK, made a note of this in the design doc